lib/code_corps/github.ex

```diff
@@ -1,41 +1,273 @@
-  Associates user with the GitHub OAuth token.
+  A low level utility function to make a direct request to the GitHub API.
+  """
+  @spec request(body, method, String.t, headers, list) :: {:ok, map} | {:error, api_error_struct}
+  def request(body, method, endpoint, headers, opts) do
+    {access_token, opts} = Keyword.pop(opts, :access_token)
+
+    base_url = get_base_url()
+    req_url = base_url <> endpoint
+    req_body = body |> Poison.encode!
+    req_headers =
+      headers
+      |> add_default_headers()
+      |> add_access_token_header(access_token)
+      |> Map.to_list()
+
+    req_opts =
+      opts
+      |> add_default_options()
+      |> add_pool_option()
 
-  Returns one of the following:
+    method
+    |> :hackney.request(req_url, req_headers, req_body, req_opts)
+    |> handle_response()
+  end
+
+  @spec build_access_token_params(String.t, String.t) :: map
+  defp build_access_token_params(code, state) do
+    @base_access_token_params
+    |> Map.put(:code, code)
+    |> Map.put(:state, state)
+  end
 
-  - {:ok, %CodeCorps.User{}}
-  - {:error, %Ecto.Changeset{}}
+  @doc """
+  A low level utility function to fetch a GitHub user's OAuth access token
   """
-  @spec associate(User.t, map) :: {:ok, User.t} | {:error, Ecto.Changeset.t}
-  def associate(user, params) do
-    user
-    |> User.github_association_changeset(params)
-    |> Repo.update()
+  @spec user_access_token_request(String.t, String.t) :: {:ok, map} | {:error, api_error_struct}
+  def user_access_token_request(code, state) do
+    req_url = get_token_url()
+    req_body = code |> build_access_token_params(state) |> Poison.encode!
+    req_headers =
+      %{"Accept" => "application/json", "Content-Type" => "application/json"}
+      |> add_default_headers()
+      |> Map.to_list()
+
+    req_opts =
+      []
+      |> add_default_options()
+      |> add_pool_option()
+
+    :hackney.request(:post, req_url, req_headers, req_body, req_opts)
+    |> handle_response()
+  end
+
+  @doc """
+  A low level utility function to make an authenticated request to the
+  GitHub API on behalf of a GitHub App or integration
+  """
+  @spec authenticated_integration_request(body, method, String.t, headers, list) :: {:ok, map} | {:error, api_error_struct}
+  def authenticated_integration_request(body, method, endpoint, headers, opts) do
+    base_url = get_base_url()
+    req_url = base_url <> endpoint
+    req_body = body |> Poison.encode!
+    req_headers =
+      headers
+      |> add_default_headers()
+      |> add_jwt_header()
+      |> Map.to_list()
+
+    req_opts =
+      opts
+      |> add_default_options()
+      |> add_pool_option()
+
+    :hackney.request(method, req_url, req_headers, req_body, req_opts)
+    |> handle_response()
+  end
+
+  @spec handle_response(http_success | http_failure) :: {:ok, map} | {:error, api_error_struct}
+  defp handle_response({:ok, status, _headers, body}) when status in 200..299 do
+    decoded_body = body |> Poison.decode!
```

Comment:
If were to delay decoding the response and instead send the encoded body out, our individual module endpoints (such as `Github.User` could automatically cast as structs:

```Elixir
defmodule Person do
  @derive [Poison.Encoder]
  defstruct [:name, :age]
end

Poison.encode!(%Person{name: "Devin Torres", age: 27})
#=> "{\"name\":\"Devin Torres\",\"age\":27}"

Poison.decode!(~s({"name": "Devin Torres", "age": 27}), as: %Person{})
```